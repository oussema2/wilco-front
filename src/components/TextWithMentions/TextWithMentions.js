import { parseValue } from 'react-native-controlled-mentions';
import { Text } from 'react-native';
import PropTypes from 'prop-types';
import React from 'react';
import { mentionStyles } from '../MentionInput/MentionInput.styles';
import AutoLink from '../AutoLink/AutoLink';
import noop from '../../helpers/noop';
import { PART_TYPE } from '../../constants/post';
import {
	isAirportMentionPartType,
	isHashtagMentionPatternPartType,
	isHashtagMentionTriggerPartType,
	isPilotMentionPartType
} from '../../helpers/mentions';

const TextWithMentions = ( {
	testID, value, style, onMentionPress, onHashtagPress
} ) => {
	const renderPart = ( part, index ) => {
		if ( isPilotMentionPartType( part ) ) {
			return (
				<Text
					testID={`${index}-${part.data?.trigger}`}
					key={`${index}-${part.data?.trigger}`}
					style={part.partType.textStyle}
					onPress={() => onMentionPress( part.data.id )}
				>
					{part.text}
				</Text>
			);
		}

		if ( isHashtagMentionTriggerPartType( part ) ) {
			return (
				<Text
					testID={`${index}-${part.data?.trigger}`}
					key={`${index}-${part.data?.trigger}`}
					style={part.partType.textStyle}
					onPress={() => onHashtagPress( part.data.name )}
				>
					{part.text}
				</Text>
			);
		}

		if ( isHashtagMentionPatternPartType( part ) ) {
			return (
				<Text
					testID={part.text}
					key={index}
					style={part.partType.textStyle}
					onPress={() => onHashtagPress( part.text?.replace( '#', '' ) )}
				>
					{part.text}
				</Text>
			);
		}

		if ( isAirportMentionPartType( part ) ) {
			return (
				<Text
					testID={part.text}
					key={index}
					style={part.partType.textStyle}
				>
					{part.text.replace( '+', '' ).toUpperCase()}
				</Text>
			);
		}

		return <AutoLink key={index} text={part.text} />;
	};

	const { parts } = parseValue( value, [
		{ trigger: PART_TYPE.mention.trigger, textStyle: mentionStyles.linkStyle },
		{ pattern: PART_TYPE.airports.pattern, textStyle: mentionStyles.airportStyle },
		{ pattern: PART_TYPE.hashtagPattern.pattern, textStyle: mentionStyles.linkStyle },
		{ trigger: PART_TYPE.hashtagTrigger.trigger, textStyle: mentionStyles.linkStyle }
	] );

	return <Text testID={testID} style={style}>{parts.map( renderPart )}</Text>;
};

TextWithMentions.propTypes = {
	testID: PropTypes.string,
	value: PropTypes.string,
	style: PropTypes.instanceOf( Object ),
	onMentionPress: PropTypes.func,
	onHashtagPress: PropTypes.func
};

TextWithMentions.defaultProps = {
	testID: 'MentionText-testID',
	value: '',
	style: {},
	onMentionPress: noop,
	onHashtagPress: noop
};

export default TextWithMentions;
