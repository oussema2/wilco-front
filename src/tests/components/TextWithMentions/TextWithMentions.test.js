import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import TextWithMentions from '../../../components/TextWithMentions/TextWithMentions';

describe( '<TextWithMentions />', () => {
	let component;
	const testID = 'testing-MentionInput-Component';
	const mockOnPressMention = jest.fn();
	const mockOnPressHashtag = jest.fn();
	const setUp = ( props ) => {
		component = render( <TextWithMentions
			testID={testID}
			onMentionPress={mockOnPressMention}
			onHashtagPress={mockOnPressHashtag}
			{...props}
		/> );
	};

	beforeEach( () => {
		setUp();
	} );

	describe( 'with the default props', () => {
		it( 'renders the TextWithMentions component correctly', () => {
			expect( component.queryByTestId( testID ) ).toBeDefined();
			expect( component ).toMatchSnapshot();
		} );
	} );

	describe( 'with no default props', () => {
		describe( 'when value has a mention', () => {
			const expectedText = '@David';

			beforeEach( () => {
				setUp( { value: 'hi @[David](1)' } );
			} );

			it( 'renders the TextWithMentions component correctly', () => {
				expect( component.queryByTestId( '1-@' ) ).toBeDefined();
				expect( component.queryByTestId( '1-@' ) ).toHaveTextContent( expectedText );
				expect( component.queryByTestId( testID ) ).toHaveTextContent( expectedText );
			} );

			describe( 'when a mention is pressed', () => {
				it( 'calls to onPress callback', () => {
					fireEvent.press( component.getByText( expectedText ) );
					expect( mockOnPressMention ).toHaveBeenCalledTimes( 1 );
				} );
			} );
		} );

		describe( 'when value has a airport', () => {
			it( 'renders the TextWithMentions component correctly', () => {
				setUp( { value: 'hi +EZE' } );
				const expectedText = 'EZE';
				expect( component.queryByTestId( '+EZE' ) ).toBeDefined();
				expect( component.queryByTestId( '+EZE' ) ).toHaveTextContent( expectedText );
				expect( component.queryByTestId( testID ) ).toHaveTextContent( expectedText );
			} );
		} );

		describe( 'when value has a hashtag', () => {
			const testHashtagPartType = ( { expectedText, customTestID, trigger = '#' } ) => {
				it( 'renders the TextWithMentions component correctly', () => {
					expect( component.queryByTestId( customTestID ) ).toBeDefined();
					expect( component.queryByTestId( customTestID ) ).toHaveTextContent( `${trigger}${expectedText}` );
				} );

				describe( 'when a hashtag is pressed', () => {
					it( 'calls to onPress callback', () => {
						fireEvent.press( component.queryByTestId( customTestID ) );
						expect( mockOnPressHashtag ).toHaveBeenCalledWith( expectedText );
					} );
				} );
			};

			beforeEach( () => {
				setUp( { value: 'hello #barcelona #[madrid](5)' } );
			} );

			describe( 'with a default hashtag', () => {
				const expectedText = 'barcelona';
				const customTestID = '#barcelona';
				testHashtagPartType( { expectedText, customTestID } );
			} );

			describe( 'with a triggered hashtag', () => {
				const expectedText = 'madrid';
				const customTestID = '3-#';
				testHashtagPartType( { expectedText, customTestID } );
			} );
		} );
	} );
} );
