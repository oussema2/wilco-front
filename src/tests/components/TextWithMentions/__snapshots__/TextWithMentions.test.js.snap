// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextWithMentions /> with the default props renders the TextWithMentions component correctly 1`] = `
<Text
  style={Object {}}
  testID="testing-MentionInput-Component"
>
  <Text
    style={Object {}}
    testID="autolink-testID"
  />
</Text>
`;
